Rebase XCP-ng changes onto v10.10.25

ydirson opened the pull request.

stormi committed:
Add software RAID support.

Originally-by: Nicolas Raynaud <github@nraynaud.com>

Also adds support for it in the answerfile.

the general form is :
<raid device="md127">
  <disk>sda</disk>
  <disk>sdb</disk>
</raid>

avoid question and display waiting message when running mdadm

Signed-off-by: Samuel Verschelde <stormi@laposte.net>

When restoring from backup on a software RAID install, install grub on
raid members instead of disk (which is a mdadm device)

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

Orig-commit: 6e168b43e0fd31cf3de9f578def1e10ba72ced35
Orig-commit: 7557a0077a65d9353a56d7a21d054ceadc29b796
Orig-commit: ee404b10b9809279600e0be121f88a675c0e2ef4
Orig-commit: e6497d1c602691104a3d750692ffceedb25c1221
Orig-commit: d33fe10fb1c36ee0326e48dbb8059a5fe42fa80b

stormi committed:
Add a netinstall boot option

The presence of this parameter changes the installer behavior,
preventing local source installation and pre-populating the xcp-ng
install url.

Originally-by: Nicolas Raynaud <github@nraynaud.com>

Use the `--netinstall` parameter that is automatically
generated at preinit stage and translate it into
`answers['netinstall'] = True`.

Signed-off-by: Samuel Verschelde <stormi@laposte.net>

Orig-commit: 36395c3c68e0d1afdd94ed8ec57d152640dde8b4
Orig-commit: 098a2f404c2d69bdd805d175943c993b88c2b4de
Orig-commit: 0b298cd20722ca0bae36ae7079df2b94491a25db

stormi committed:
Import XCP-ng GPG key to yum and rpm post-install

Signed-off-by: Samuel Verschelde <stormi@laposte.net>

Orig-commit: 9f0f094409a7f0c19a686cad3ed16f055cbbf155
Orig-commit: eb796dc6a82a0defdcbba1a0833a3356060c5396

stormi committed:
Use xcp-ng-deps instead of groups.xml

Signed-off-by: Samuel Verschelde <stormi@laposte.net>

Orig-commit: 2731116cba1ef80a2628c041002f82258e1be5cb

stormi committed:
kernel-alt support

- show kernel-alt warning

Originally-by: Rushikesh Jadhav <rushikesh7@gmail.com>

- Update warning message when installing with kernel-alt
- cleaner reboot action on that screen: return EXIT
- fix detection of the kernel-alt boot parameter of the installer
- install kernel-alt when booting on install-alt
- create initrd and update grub configuration as post-install task

Signed-off-by: Samuel Verschelde <stormi@laposte.net>

Orig-commit: 243ce81f0ff4ad17c99c1ca6af5f3326af1c94db
Orig-commit: d9a8a3eb609b6c0493fb3b6db03fc91bc5321f45
Orig-commit: 7d9e1b0144beea1de948fe3f6b36b82fa898196e
Orig-commit: 94479647318a4e04dd2ab876c075284d142809eb

ydirson committed:
Add support for globally disabling repo_gpgcheck

While developing and/or installing from an already-verified ISO, yum's
repo_gpgcheck just gets in the way.  This provides a general kill-switch
for checking repomd.xml.asc, through commandline or answerfile.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
Add support for globally disabling gpgcheck

Similar to no-repo-gpgcheck but for RPM sigs.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
answerfile: support *gpgcheck override for each <source>

Global *gpgcheck flag can change the default from True to False, and these
new flags allow to override this default on a per-source basis.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
Remove mention of Technical Support Representative in error messages

Originally-by: Samuel Verschelde <stormi@laposte.net>
Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

Orig-commit: b20e3033ecc1072f01f1b4cf619f3b2b952e592f

ydirson committed:
Merge branch '8.2.1-small-patches'

Rebase of simple 8.2.1 patches onto 8.3 codebase

ydirson committed:
Fix interactive install to honor no-*gpgcheck from commandline

When originally implementing the per-source gpgcheck flags in
answerfile[1], the full code was moved to an anwerfile-only location,
breaking the original interactive-install implementation (and then
following code review[2] the working code for interactive install
disappeared further from the patch series).

This moves the Repository flag-setting to `add_repos()` common code,
while leaving the flag computation to the caller, since only the
answerfile case has to do any non-trivial logic.

- [1] https://github.com/xcp-ng/host-installer/commit/06b700789f2c22d1c78285f9583937511e4d3c07
- [2] https://github.com/xcp-ng/host-installer/pull/2#discussion_r1012080594

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

stormi committed:
Update URL for netinstall in 8.3

Signed-off-by: Samuel Verschelde <stormi-xcp@ylix.fr>

ydirson committed:
RAID support: zap partition tables of disks before building a RAID

Failing to do so can leave the kernel's idea of the partitions in weird
states.  There are likely bugs underlying those weird state, we just don't
want to be impacted by them at this point :)

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
Merge branch 'lvm-vs-ext'

ydirson committed:
Local SR: make EXT the default and first option instead of LVM

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
RAID: only propose creation with at least 2 disks and no RAID

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

benjamreis committed:
Write `manual_nameservers` and `domain` for IPv6 too

Both fields are used for the 2 IP families and were writen in
an IPv4 specific block.

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Add `NetInterfaceV6` to init an IPv6 interface

Inherits from `NetInterface` to mutualize the code

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
isStatic helpers by IP family

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Use `socket.inet_pton` to validate IP

Add family specific validator as well

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Take IPv6 addresses into account to determine an interface is up

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Enable IPv6 in the host when it is configured

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Write IPv6 conf files for the installer

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Configure IPv6 in TUI

Add a screen to choose to configure IPv4, IPv6 or both
Display the interface screen for IPv4 and/or IPv6

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
NetInterface inherits from Object

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
`waitUntilUp` also wait for IPv6 if statically configured

Need to add `ndisc6` to the install.img

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Hide dynamic buttons in tui when in static mode

Add `isDynamic` helper to `NetInterface`

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis committed:
Keep IPv6 enablement/disablement upon upgrades

Add `etc/sysctl.d/91-net-ipv6.conf` in the restore list

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

stormi committed:
Merge pull request #13 from ipv6

IPv6 support

benjamreis committed:
Keep user multipath configuration upon upgrade

All files matchin `custom.*\.conf` will be kept
upon upgrade.

See: https://github.com/xapi-project/sm/pull/600

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

ydirson committed:
getDiskDeviceSize: don't silently return None (part of #66)

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
XenServerBackup: ignore backups with inaccessible primary disk (#66)

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
findXenSourceBackups: refactor and log reason for ignoring

Previous code structure was that of 2 nested if's, with nominal code
path being in the positive branch of each if.  Adding another
condition will bring a need for logging the reason to ignore a backup,
but doing that by converting inner `if` to an `elif` chain (arguably
the simplest modification) would bring the nominal code path to switch
to the negative/last branch of the inner `elif` chain.  At the same
time, the outer `if` would not seem special enough to deserve it
special place (and the cyclomatic complexity).

This commit leverages the existing `try:except:` block to switch to an
"error out the loop" pattern, where the nominal code path is just
linear.

Using `StopIteration` may feel like an abuse, but:
- it is the only standard `Exception` that is not a `StandardError` or
  a `Warning`, and defining a new one could seem overkill
- the closest alternative would be a `while True` loop and breaking out
  in both exceptional code paths and at the end of nominal path, where
  the `break` statements would "stop the iteration" as well

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
findXenSourceBackups: log any exception caught

- it is bad practice to "catch any"
- not logging anything just hides information, which can be especially
  important here as the reason for a try/catch is not obvious (exceptions
  thrown by XenServerBackup.__init__?)

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
installFromYum: give more detailed error messages on gpg errors

Covers:
1. repo_gpgcheck:
  a. wrong system clock putting gpg key creation in the future, causing a
     yum crash (nothing special happens if the date of the signature is in
     the future ¯\_(ツ)_/¯)
  b. other yum crashes due to uncaught gpg exceptions (if any)
  c. lack of repomd signature (while repo_gpgcheck is in force)
  d. signature done by other key than the one in ISO ("repomd.xml signature
     could not be verified" ¯\_(ツ)_/¯)
2. gpgcheck:
  a. RPM signed with unknown key
  b. unsigned RPM referenced by unsigned repomd (no-repo-gpgcheck)
  c. RPM re-signed with unknown key, unsigned repomd (no-repo-gpgcheck)
  d. RPM overwritten with another RPM signed with known key (diagnosed
     through hash but, same diag as 2.c)
  e. delsigned/resigned/etc RPM, unchanged repomd (same diag as 2.c/d)

Does not cover notably:
  - unsigned RPM referenced by (re)signed repomd

In some cases Yum does not give an error, but dies because of an
uncaught exception, which makes this check quite brittle, but in the
worst case if messages change, we still fallback to the original
"Error installing packages" message.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
Merge pull request #20 from xcp-ng/inexistant-disks

Don't fail seeing a backup refering to non-existent disk

ydirson committed:
Merge pull request #21 from xcp-ng/gpg-diags

installFromYum: give more detailed error messages on gpg errors

ydirson committed:
Merge branch 'restore-multipath-conf' into 10.10.9-8.3

* xcpng/restore-multipath-conf:
  Keep user multipath configuration upon upgrade

ydirson committed:
Always show Supplemental Packs dialog if Driver Disks are used.

In interactive mode, when Driver Disks are used as local media, they
must *also* be installed as Supplemental Packs after installation.
This change remembers the use of Driver Disks and makes sure the
question is asked, even when the supplemental-packs feature is
disabled.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
Clarify Supplemental Packs dialog in presence of Driver Disks

It may not be obvious to users that local-media Driver Disks must be
inserted twice during installation.  This change uses the memorization
of usage of Driver Disks from previous commit, and uses it to explain
users they likely wants to insert their Driver Disk again.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
answerfile: change installation's sr-type default to "ext"

This complements the default choice previously changed in the TUI.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

rosslagerwall committed:
disktool: Optionally dump old/new partitions during commit

If log=True is passed to commit, dump the old and new partition tables
before applying the changes. At the same time, fix up the dump code to
handle GPT partition tables.

Signed-off-by: Ross Lagerwall <ross.lagerwall@citrix.com>
(cherry picked from commit 7f5877188337915091738e88222593488b0e8039)

liulinC committed:
CP-47621: Enable ability to set hostname from DHCP

According to https://www.freedesktop.org/software/systemd/man/latest/hostname.html
Systemd set hostname with following sequence
- kernel parameter, systemd.hostname
- static hostname in /etc/hostname
- transient hostname like DHCP
- localhost at systemd compile time

Set static hostname `localhost.localdomain` in /etc/hostname
prevents obtain hostname from DHCP, and all XenRT installed host
got localhost.localdomain as the hostname

This commit set empty static hostname to permit DHCP

Signed-off-by: Lin Liu <lin.liu@citrix.com>
(cherry picked from commit e39f2686df1c669a0427d4c2834f9db8ecec97d3)

ydirson committed:
Local SR: use a RadioBar for LVM/EXT choice

Also change text for a more suitable description.

Originally-by: Samuel Verschelde <stormi@laposte.net>
Orig-commit: b4ebf8174694799b19eaa5a946ba7f00827eeef5
Orig-commit: 03d3a84ce27976a1d9d6ea9e42a1e7492967f3e6
Orig-commit: d3857b8f23e7e02800e5a531a01b184482023479
Orig-commit: 604f0b07aa090ddf5ccbf1c2793d16a7794dec2f

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
__mkinitrd: pass -f to dracut

Before 5abd998733a76f3609ca1394e4d923eda8f1a8a3 this code was calling
grubby's new-kernel-pkg, which unconditionally passes -f to mkinitrd.
Not using -f will cause install failure when a driver disk's rpm's
scripts have generated an initrd already (as is the case with the
intel igb driver)

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
Merge pull request #32 from xcp-ng/for-xs8/fix-driver-disk

__mkinitrd: pass -f to dracut

ydirson committed:
Enhance tracing to hunt for restore bug

The particular issue turned out to be that the nested UEFI host being
restored had several disks attached.  Since this is done as NVMe namespaces
of a single disk, all sharing the same ID, udev in dom0 does not have
enough info and messes up /dev/disk/by-id/ on which restore relies because
of /etc/xensource-inventory:

 lrwxrwxrwx 1 root root 13 Jul 15 14:28 nvme-QEMU_NVMe_Ctrl_nvme0 -> ../../nvme0n3
 lrwxrwxrwx 1 root root 15 Jul 15 14:28 nvme-QEMU_NVMe_Ctrl_nvme0-part1 -> ../../nvme0n2p1
 lrwxrwxrwx 1 root root 15 Jul 15 14:28 nvme-QEMU_NVMe_Ctrl_nvme0-part2 -> ../../nvme0n1p2

Only the addition of the `tool.dump()` call allowed me to connect the
"Failed to create root filesystem" error with that longstanding (in)famous
issue, but a few other additions were useful to get there.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>
(cherry picked from commit 02d72e663450aa478a1c7087191419e856988a01, with
one additional safety guard that does not apply to master any more)

benjamreis committed:
Mark 'Dual Stack' as experimental

Signed-off-by: Benjamin Reis <benjamin.reis@vates.tech>

ydirson committed:
Merge pull request #34 from xcp-ng/xs8/enhance-traces

Enhance tracing to hunt for restore bug

ydirson committed:
Merge pull request #33 from xcp-ng/xs8/accept-gpt-partid

disktool: Optionally dump old/new partitions during commit

tescande committed:
Ignore backup if boot modes don't match

This patch checks the boot mode of an existing backups and disallows its
restoration if it differs from the installer one. This uses the new
helper function is_rootf_uefi().

Signed-off-by: Thierry Escande <thierry.escande@vates.tech>

tescande committed:
Ignore existing installation for upgrade if boot modes don't match

This patch checks that the current installation is not proposed as
upgradeable if its boot mode differs from the installer one. This uses
the helper function is_rootfs_uefi().

Signed-off-by: Thierry Escande <thierry.escande@vates.tech>

tescande committed:
Add helper function is_rootfs_uefi()

This function searches for the /boot/efi mount point in the /etc/fstab
file of the mount point directory passed as parameter and returns True
if present, False otherwise.

This will allow to check that an existing installation or backup boot
mode matches with the installer boot mode.

Signed-off-by: Thierry Escande <thierry.escande@vates.tech>

ydirson committed:
Upgrade: forbid upgrading with a key XAPI will reject

XAPI now rejects the default keysize of 7.x era, which must be
regenerated before upgrading to 8.3.  Let the installer refuse to
initiate a situation where a Rolling Pool Upgrade would be unable to
proceed, with not-yet-updated slaves holding the running VMs getting
refused connection to the updated part of the pool.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
Upgrade: send users to XCP-ng 8.3 relnotes for details.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
Merge pull request #40 from xcp-ng/certificate-keysize-protection

Upgrade: forbid upgrading with a key XAPI will reject

ydirson committed:
Merge branch 'install-on-raid1'

ydirson committed:
Merge branch 'netinstall'

ydirson committed:
Merge pull request #2 from xcp-ng/yann/gpgcheck

Global and per-source gpgcheck flags

ydirson committed:
Merge branch 'driver-disks'

* driver-disks:
  Clarify Supplemental Packs dialog in presence of Driver Disks
  Always show Supplemental Packs dialog if Driver Disks are used.

ydirson committed:
Merge branch 'boot-mode-checks' into 10.10.19-8.3

stormi commented:
Oh no, one of these reviews :D

stormi commented:
Have we documented review tips last time? If not, that would be a nice addition to the docs. Steps to compare with the previous branch, via exported patches or smart git fu.

ydirson committed:
Makefile: give more flexibility to select base multipath.conf

Useful to take that file from a source tree rather than a RPM
deployed in /.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>
(cherry picked from commit 96db5f422688ca88899eef812897638352e478c8)
(cherry picked from commit 3c6b1ff3c3c5cc50baba7b3dcb5dc685e244ad7d)

ydirson commented:
This is a good start:
```
git range-diff v10.10.19..10.10.19-8.3 v10.10.25..ydi/10.10.25-8.3
```

ydirson committed:
Merge branch 'dhcp-hostname' (PR #196)

Backport from master.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson commented:
Last push:
- avoid successive merges of ipv6 branch
- moved a logging improvement to the commit it belongs (was problematically in a merge)
- added a backport of https://github.com/xenserver/host-installer/pull/215 to `small-fixes` branch, for easier testing (see https://github.com/xcp-ng/xcp/pull/684)

ydirson committed:
Avoid attempting to re-read inventory file from unmounted partition

e9d686ae17bfcad5383f2d89d6bf848c3abd9a3b triggers an unmount of backup_fs
before we're done with it, causing a restoration failure in 10.10.25:

 INFO     [2025-03-19 14:33:50] INSTALL FAILED.
 INFO     [2025-03-19 14:33:50] A fatal exception occurred:
 INFO     [2025-03-19 14:33:50] Traceback (most recent call last):
   File "/opt/xensource/installer/install.py", line 274, in go
     restore.restoreFromBackup(backup, progress)
   File "/opt/xensource/installer/restore.py", line 214, in restoreFromBackup
     branding = util.readKeyValueFile(os.path.join(backup_fs.mount_point, constants.INVENTORY_FILE))
   File "/opt/xensource/installer/util.py", line 322, in readKeyValueFile
     f = open(filename, "r")
 IOError: [Errno 2] No such file or directory: '/tmp/restore-backup-hrJwAR/etc/xensource-inventory'


Fortunately that data we want was already loaded, so this is a minimal fix.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
Merge branch 'fix-restore' into 10.10.25-8.3

See https://github.com/xenserver/host-installer/pull/220

glehmann commented:
```
❯ git range-diff v10.10.19..10.10.19-8.3 v10.10.25..ydi/10.10.25-8.3
fatal : bad revision 'v10.10.25..ydi/10.10.25-8.3'
```
because the tags have not been imported from upstream.
~~~
git remote add upstream git@github.com:xenserver/host-installer.git
git fetch --tags upstream
~~~
fixes that

ydirson commented:
> the tags have not been imported from upstream.

Pushed `v10.10.25` here

stormi reviewed (approved):
Trusting Yann's work and Gaëtan's review.

ydirson commented on upgrade.py:
```diff
@@ -666,6 +681,12 @@ def buildRestoreList(self, src_base):
 
         return restore_list
 
+        # Keep IPv6 enablement/disablement upon upgrades
+        self.restore_list += ['etc/sysctl.d/91-net-ipv6.conf']
+
+        # Keep user multipath configuration
+        self.restore_list += [{'dir': 'etc/multipath/conf.d', 're': re.compile(r'custom.*\.conf')}]
+
```
Wow, @glehmann somehow we missed this mismerge

ydirson replied:
Indeed `git range-diff` ignores merge commits and does not give the option to include them.

ydirson replied:
we'd best rebase the ipv6 branch on 10.10.25+ instead to reduce conflicts
